feat: automatically upgrade stream types to their latest definitions

doudou opened the pull request:
Depends on:
- [ ] https://github.com/rock-core/tools-pocolog/pull/41

On top of #43 

Fully automatic for replay (since it is a hard requirement) and can be disabled for file export. Not implemented for plain data analysis.

doudou committed:
feat: optionally store a reference of a dataset's datastore

doudou committed:
feat: add basic support for an optional datastore-wide config file

doudou committed:
feat: load a pocolog upgrade converter registry from a datastore-wide config path

doudou committed:
feat: implement value-level upgrade method in Dataset

This implementation allows to upgrade values from "old" types to their
new representations, as well as upgrade types from different layouts,
e.g. from a ARM32 to a x86-64 architecture to pass to e.g. C++.

doudou committed:
feat: auto-upgrade log data during replay

doudou committed:
feat: implement data upgrade in export_to_single_file

wvmcastro commented on test/datastore_test.rb:
```diff
@@ -128,6 +128,67 @@ module Syskit::Log
+            it "reads the upgrade handlers path" do
+                path = @datastore_path / "upgrades"
+                config = { "upgrade_handlers_path" => path.to_s }
+                (@datastore_path / "config.yml").write(YAML.dump(config))
```
```suggestion
                apply_datastore_config(config)
```

wvmcastro commented on lib/syskit/log/datastore/dataset.rb:
```diff
@@ -775,6 +776,98 @@ def each_task(
+            #
+            # For more complex upgrades, one has to provide custom conversion handlers. These
+            # handlers are Ruby files that register conversions on a
+            # {Pocolog::Upgrade::ConversionRegistry} object. See the rock-and-syskit
+            # documentation for more details
```
Documentation yet to come?

doudou replied:
Yes

wvmcastro commented on lib/syskit/log/datastore/dataset.rb:
```diff
@@ -775,6 +776,98 @@ def each_task(
+
+                @upgrade_ops_for_target[cache_key] ||=
+                    Pocolog::Upgrade.compute(
+                        reference_time, value.class, target_type,
```
Why Pocolog's upgrade graph needs a reference time? How does this impact how a conversion from type A to B is done?

doudou replied:
This is for conversions from type A to type A, but fixing some issues in the data itself. Generally speaking, all converters have a "up to" date that limits their scope. Since converters are written in Ruby, it can be Time.now if said converter has no obvious due date.

doudou commented:
Ping @wvmcastro

doudou committed:
fix: usage of Streams for plain log files and datasets

The introduction of the auto-upgrade feature broke this use-case,
which is heavily used in tests (at least). Making it to function
again is not a lot of work, keep the feature for now even if we
don't actually use it.

doudou committed:
chore: style

doudou committed:
chore: style
